### ISSUE TITLE
Concurrent staticcheck calls eating RAM

### ISSUE THREAD

**Issue by fasaxc** [alecthomas/gometalinter#273]
Each call to staticcheck on my repo seems to use 1.5GB of RAM and gometalinter runs several in parallel, which quickly results in swapping on my machine.  OTOH, staticcheck seems to support running against multiple directories and that seems to use a sensible amount of RAM.

**Comment by alecthomas** [comment]
You would think this would be simply a matter of passing `...` paths to the various linters. I have a change that fixes this, but staticcheck (among others) include vendor paths when left to their own devices. Not only does this result in lint output for vendor code that we really don't care about, but on large vendor directories you're checking a lot of code for no reason. Alternatively, some linters support passing a list of directories to check, so gometalinter could (and does) expand `...` itself, excluding vendor, but some linters don't support this.

Basically, it's not as simple as one would hope.

I have a CL that does the path expansion, but it does not work in all cases. I have yet to get around to fixing it.

**Comment by fasaxc** [comment]
@alecthomas FWIW, it looks like staticcheck does support passing multiple paths

**Comment by alecthomas** [comment]
Yeah it does, but other linters don't, or don't behave correctly when you do. So gometalinter can't simply apply that logic to all linters.

**Comment by bradleyfalzon** [comment]
For the vendor paths, staticcheck (and other linters) use gotool for package expansion, and since https://github.com/golang/go/issues/19090 was accepted gotool will likely add support for `...` to exclude vendor, for this too see https://github.com/kisielk/gotool/issues/16.

Would this help at all?

**Comment by alecthomas** [comment]
I imagine this would very much help, yes.

**Comment by alecthomas** [comment]
I just realised that this is probably due to gometalinter disabling GC. This normally results in not-insignificant speedups. You can try running with `--enable-gc`.

**Comment by pierrre** [comment]
@alecthomas 
FYI I was running `--enable-gc` until recently, because gometalinter was using too much RAM.
Sadly, my projects grew bigger and bigger, and gometalinter was slower: more than 10 minutes on a single project.
After removing `--enable-gc`, gometalinter was 2x faster.

I've tried to tweak the `GOGC` variable, while using `--enable-gc`. (gometalinter forwards the env variable to children processes)
Increasing the value to `400` or `800` speeds up the linters, while keeping RAM usage reasonable.
I know that tweaking this variable is not recommended, but it works.

My proposal:
- remove `--enable-gc` (or just disable it)
- add a new flag `--gc`, which will set the `GOGC` variable, with a reasonable default value (maybe 400 or 800)

If the `GOGC` variable is defined before running gometalinter, it overrides the default value.
If `--gc` is defined, it overrides the `GOGC` variable.

What do you think ?

**Comment by alecthomas** [comment]
This should (in theory) be largely fixed by #289. Can you try out that PR?

**Comment by alecthomas** [comment]
But it does seem reasonable to set `GOGC=800` rather than disabling it completely. It might be sufficient to just change `GOGC=off` to `GOGC=800`...

**Comment by alecthomas** [comment]
This should be fixed in master.
